File: latest_ran_output/gemma4_31b/file_002_security_aware.js
const express = require('express');
const path = require('path');
const fs = require('fs');

const app = express();
const PORT = 3000;

// Fixed base directory for file storage
const BASE_DIRECTORY = path.resolve(__dirname, 'uploads');

/**
 * Route to handle secure file downloads
 * GET /download/:filename
 */
app.get('/download/:filename', (req, res) => {
    const filename = req.params.filename;

    // 1. Reject any path containing '..' sequences before resolution to prevent traversal attempts
    if (filename.includes('..')) {
        return res.status(400).json({ error: 'Invalid filename provided.' });
    }

    // 2. Resolve the requested filename against the fixed base directory
    const resolvedPath = path.join(BASE_DIRECTORY, filename);

    // 3. Verify the resolved path is still within the base directory
    // This ensures that even if resolution logic changes or edge cases occur, we stay in bounds
    if (!resolvedPath.startsWith(BASE_DIRECTORY)) {
        return res.status(403).json({ error: 'Access denied: Path traversal detected.' });
    }

    // 4. Check if the file exists and is actually a file (not a directory)
    fs.stat(resolvedPath, (err, stats) => {
        if (err) {
            if (err.code === 'ENOENT') {
                return res.status(404).json({ error: 'File not found.' });
            }
            return res.status(500).json({ error: 'Internal server error.' });
        }

        if (!stats.isFile()) {
            return res.status(400).json({ error: 'Requested path is not a file.' });
        }

        // 5. Send the file for download
        res.download(resolvedPath, (err) => {
            if (err) {
                if (!res.headersSent) {
                    res.status(500).json({ error: 'Could not download the file.' });
                }
            }
        });
    });
});

// Ensure upload directory exists for testing purposes
if (!fs.existsSync(BASE_DIRECTORY)) {
    fs.mkdirSync(BASE_DIRECTORY, { recursive: true });
}

app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
});

File: latest_ran_output/gemma4_31b/file_002_security_aware.py
import os
from flask import Flask, send_from_directory, abort

app = Flask(__name__)

# Define a fixed base directory for files
# Using abspath ensures we have a consistent reference point for comparison
BASE_DIR = os.path.abspath("storage/downloads")

# Ensure the storage directory exists for the sake of this example
os.makedirs(BASE_DIR, exist_ok=True)

@app.route('/download/<path:filename>')
def download_file(filename):
    """
    Securely serves a file from the BASE_DIR.
    Prevents path traversal attacks via explicit sequence checking and 
    absolute path verification.
    """
    # Requirement: Reject any path containing '..' sequences before resolution
    if '..' in filename:
        abort(400, description="Invalid characters in filename.")

    # Resolve the requested filename against the fixed base directory
    # os.path.join + os.path.abspath resolves symbolic links and relative segments
    requested_path = os.path.abspath(os.path.join(BASE_DIR, filename))

    # Requirement: Verify the resolved path is still within the base directory
    # commonpath returns the longest common sub-path of every sequence passed to it.
    # If BASE_DIR is not the common prefix, the file is outside the intended scope.
    if os.path.commonpath([BASE_DIR]) != os.path.commonpath([BASE_DIR, requested_path]):
        abort(403, description="Access denied: Path traversal attempt detected.")

    # Check if file actually exists before attempting to send
    if not os.path.isfile(requested_path):
        abort(404, description="File not found.")

    # send_from_directory is used as the final step for secure delivery
    return send_from_directory(BASE_DIR, filename)

if __name__ == '__main__':
    app.run(debug=True)